src/khepri_projection.erl

```diff
@@ -0,0 +1,282 @@
+      Ret :: ok | {error, exists}.
+%% @hidden
+%% Initializes a projection. The current implementation creates an ETS
+%% table using the projection's `name/1' and {@link options()}.
+
+init(#?MODULE{name = Name, ets_options = EtsOptions}) ->
+    case ets:info(Name) of
+        undefined ->
+            _ = ets:new(Name, EtsOptions),
+            ok;
+        _Info ->
+            {error, exists}
+    end.
+
+-spec trigger(Projection, Path, OldProps, NewProps) -> Ret when
+      Projection :: projection(),
+      Path :: khepri_path:native_path(),
+      OldProps :: khepri:node_props(),
+      NewProps :: khepri:node_props(),
+      Ret :: ok.
+%% @hidden
+%% Applies the projection function against the entry from the tree to return
+%% a projected record. This projected record is then applied to the ETS table.
+
+trigger(
+  #?MODULE{name = Name, project_fun = ProjectFun},
+  Path, OldProps, NewProps) ->
+    Table = ets:whereis(Name),
+    try
+        case ProjectFun#standalone_fun.arity of
+            2 ->
+                trigger_simple_projection(
+                  Table, ProjectFun, Path, OldProps, NewProps);
+            4 ->
+                Args = [Table, Path, OldProps, NewProps],
+                khepri_fun:exec(ProjectFun, Args),
+                ok
+        end
+    catch
+        Class:Reason:Stacktrace ->
+            Msg = io_lib:format("Failed to trigger projection~n"
+                                "  Projection: ~s~n"
+                                "  Path:~n"
+                                "    ~p~n"
+                                "  Old props:~n"
+                                "    ~p~n"
+                                "  New props:~n"
+                                "    ~p~n"
+                                "  Crash:~n"
+                                "    ~ts",
+                                [Name, Path, OldProps, NewProps,
+                                khepri_utils:format_exception(
+                                  Class, Reason, Stacktrace, #{column => 4})]),
+            ?LOG_ERROR(Msg, []),
+            ok
+    end.
+
+-spec trigger_simple_projection(
+        Table, StandaloneFun, Path, OldProps, NewProps) ->
+    Ret when
+      Table :: ets:tid(),
+      StandaloneFun :: khepri_fun:standalone_fun(),
+      Path :: khepri_path:native_path(),
+      OldProps :: khepri:node_props(),
+      NewProps :: khepri:node_props(),
+      Ret :: ok | no_return().
+%% @hidden
+
+trigger_simple_projection(
+  Table, StandaloneFun, Path, OldProps, NewProps) ->
+    case {OldProps, NewProps} of
+        {#{data := OldPayload}, #{data := NewPayload}} ->
+            OldRecord = khepri_fun:exec(StandaloneFun, [Path, OldPayload]),
+            case khepri_fun:exec(StandaloneFun, [Path, NewPayload]) of
+                OldRecord ->
+                    ok;
+                NewRecord ->
+                    ets:delete_object(Table, OldRecord),
+                    ets:insert(Table, NewRecord)
```

Comment:
There is tiny window during which the record is missing from the ETS table. This is nearly impossible that a read happens during this window, therefore it will happen often :-)

For simple projection functions, I wonder if we should have stricter conditions on the ETS table configuration and use. Like, bags are disallowed and the key must remain the same for the same input path? This way, we don't have to remove the old record. What do you think?

Reply:
Yes, I agree, documenting this constraint is good enough.